x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/ScoreScriptUtils.java

```diff
@@ -30,10 +30,34 @@
         final float[] queryVector;
 
         public DenseVectorFunction(ScoreScript scoreScript, List<Number> queryVector) {
+            this(scoreScript, queryVector, false);
+        }
+
+        /**
+         * Constructs a dense vector function.
+         *
+         * @param scoreScript The script in which this function was referenced.
+         * @param queryVector The query vector.
+         * @param normalizeQuery Whether the provided query should be normalized to unit length.
+         */
+        public DenseVectorFunction(ScoreScript scoreScript,
+                                   List<Number> queryVector,
+                                   boolean normalizeQuery) {
             this.scoreScript = scoreScript;
+
             this.queryVector = new float[queryVector.size()];
+            double queryMagnitude = 0.0;
             for (int i = 0; i < queryVector.size(); i++) {
-                this.queryVector[i] = queryVector.get(i).floatValue();
+                float value = queryVector.get(i).floatValue();
+                this.queryVector[i] = value;
+                queryMagnitude += value * value;
+            }
+            queryMagnitude = Math.sqrt(queryMagnitude);
+
+            if (normalizeQuery) {
```

Comment:
may be worth to put sqrt calculation (line 55) under if(normalizeQuery) condition as well